### PR TITLE
Catch invalidated key

### PR DESCRIPTION
**Pull request by GavinPacini**
This is my proposed solution. This is only for the `1.x` branch as that is what I am currently using but should be easy to port to `2.x`.

### CODE REVIEW

File: rxfingerprint/src/main/java/com/mtramin/rxfingerprint/FingerprintObservable.java [hunk 1]
```diff
@@ -115,6 +117,20 @@ public void onAuthenticationSucceeded(FingerprintManagerCompat.AuthenticationRes
 	}
 
 	/**
+	 * Required to catch errors from certain phones. See:
+	 * http://stackoverflow.com/questions/36043912/error-after-fingerprint-touched-on-samsung-phones-android-security-keystoreexce
+	 * @param e Exception caught
+	 * @param subscriber Subscriber for fingerprint result
+     */
+	void checkKey(Exception e, Subscriber<? extends FingerprintAuthenticationResult> subscriber, Action0 action0) {
```
Comment (Mauin) [inline rxfingerprint/src/main/java/com/mtramin/rxfingerprint/FingerprintObservable.java]: This feels like it solves a completely different issue for some specific devices. What does this have to do with the invalidation of the key that `RxFingerprint` uses internally as discussed in #24 ?

File: rxfingerprint/src/main/java/com/mtramin/rxfingerprint/FingerprintEncryptionObservable.java [hunk 2]
```diff
@@ -115,7 +128,7 @@ protected void onAuthenticationSucceeded(Subscriber<FingerprintEncryptionResult>
 			subscriber.onNext(new FingerprintEncryptionResult(FingerprintResult.AUTHENTICATED, null, encryptedString));
 			subscriber.onCompleted();
 		} catch (CryptoDataException | IllegalBlockSizeException | BadPaddingException | InvalidParameterSpecException | UnsupportedEncodingException e) {
-			subscriber.onError(e);
+			checkKey(e, subscriber, () -> subscriber.onNext(new FingerprintEncryptionResult(FingerprintResult.INVALIDATED, null, null)));
```
Comment (Mauin) [inline rxfingerprint/src/main/java/com/mtramin/rxfingerprint/FingerprintEncryptionObservable.java]: Why does encryption return a different result from this error case than decryption? Here we return `FingerprintResult.INVALIDATED` whereas decryption still throws `KeyPermanentlyInvalidatedException`.

Comment (GavinPacini) [inline rxfingerprint/src/main/java/com/mtramin/rxfingerprint/FingerprintEncryptionObservable.java]: My thinking here was that if you are encrypting, it could be expected behaviour (hence using `onNext`). However, if decrypting, the data is lost so it should be an error (hence `onError`).

I agree it might be unclear. Do you think from an API standpoint it would be better to just use `FingerprintResult.INVALIDATED` in both cases?

File: rxfingerprint/src/main/java/com/mtramin/rxfingerprint/CryptoProvider.java [hunk 3]
```diff
@@ -140,4 +140,11 @@ private SecretKey getKey(String keyName) throws KeyStoreException, CertificateEx
         keyStore.load(null);
         return (SecretKey) keyStore.getKey(keyName, null);
     }
+
+    @TargetApi(Build.VERSION_CODES.M)
```
Comment (GavinPacini) [inline rxfingerprint/src/main/java/com/mtramin/rxfingerprint/CryptoProvider.java]: Will remove this now.

File: rxfingerprint/src/main/java/com/mtramin/rxfingerprint/FingerprintObservable.java (hunk 1 above)
Comment (GavinPacini) [inline rxfingerprint/src/main/java/com/mtramin/rxfingerprint/FingerprintObservable.java]: Well, the issue is that the `KeyPermanentlyInvalidatedException` can (on some popular devices) manifest itself as a `android.security.KeyStoreException`. So, while it may not seem like the same issue, if you read the referenced stackoverflow link, it is caused by a fingerprint being added while the app is in the background.

So, I felt it was necessary to resolve both rather than just one. :)

Comment (Mauin) [inline rxfingerprint/src/main/java/com/mtramin/rxfingerprint/FingerprintObservable.java]: While that might be the case, and certainly an issue worth fixing it's not addressing the issue as described in #24. I'd therefore handle these cases in another issue/PR.

File: rxfingerprint/src/main/java/com/mtramin/rxfingerprint/FingerprintEncryptionObservable.java (hunk 2 above)
Comment (Mauin) [inline rxfingerprint/src/main/java/com/mtramin/rxfingerprint/FingerprintEncryptionObservable.java]: I'm unsure if the `FingerprintResult.INVALIDATED` gives the user of the library any additional information.

Once the key is invalidated this will be notified by the `onError` callback containing the `KeyPermanentlyInvalidatedException`. Consumers should then clear their stored data and it we should be able to simply remove/recreate the key if we receive `KeyPermanentlyInvalidatedException` in the `encrypt` call.

For that case I don't see the added benefit from having an additional `FingerprintResult`. What do you think?

**Comment by Mauin** [comment]
Closing via #34. Thanks for the contribution! 👍
